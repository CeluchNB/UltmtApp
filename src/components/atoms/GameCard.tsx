import { Game } from '../../types/game'
import Icon from 'react-native-vector-icons/MaterialCommunityIcons'
import { IconButton } from 'react-native-paper'
import { useTheme } from '../../hooks'
import { Animated, Pressable, StyleSheet, Text, View } from 'react-native'
import React, { useEffect } from 'react'

export interface GameCardProps {
    game: Game
    onPress: (id: string) => void
}

const GameCard: React.FC<GameCardProps> = props => {
    const {
        theme: { colors },
    } = useTheme()
    const liveOpacity = React.useRef(new Animated.Value(0)).current

    const { game, onPress } = props
    const {
        _id,
        teamOne,
        teamTwo,
        teamOneScore,
        teamTwoScore,
        scoreLimit,
        teamOneActive,
        totalViews,
    } = game

    useEffect(() => {
        Animated.loop(
            Animated.sequence([
                Animated.timing(liveOpacity, {
                    toValue: 1,
                    duration: 750,
                    useNativeDriver: true,
                }),
                Animated.timing(liveOpacity, {
                    toValue: 0,
                    duration: 750,
                    useNativeDriver: true,
                }),
            ]),
        ).start()
    }, [liveOpacity])

    const styles = StyleSheet.create({
        container: {
            backgroundColor: colors.primary,
            marginTop: 10,
            marginRight: 5,
            marginLeft: 5,
            color: colors.textPrimary,
            borderColor: colors.textPrimary,
            borderWidth: 1,
            borderRadius: 8,
            elevation: 10,
            shadowColor: colors.textPrimary,
        },
        teamContainer: {
            display: 'flex',
            flexDirection: 'row',
            paddingLeft: 5,
            paddingRight: 5,
            paddingTop: 5,
            color: colors.textPrimary,
        },
        teamNameContainer: {
            flex: 1,
        },
        teamNameText: {
            color: colors.textPrimary,
        },
        teamText: {
            color: colors.textSecondary,
            fontWeight: 'bold',
            fontSize: 20,
        },
        footer: {
            display: 'flex',
            flexDirection: 'row',
        },
        footerText: {
            alignSelf: 'center',
            fontSize: 16,
            fontWeight: '600',
            color: colors.textPrimary,
        },
        scoreLimitContainer: {
            display: 'flex',
            flexDirection: 'row',
            alignSelf: 'flex-end',
        },
        viewsContainer: {
            display: 'flex',
            flexDirection: 'row',
            alignItems: 'center',
            flex: 1,
        },
        viewsMargin: {
            margin: 5,
        },
        circle: {
            width: 10,
            height: 10,
            borderRadius: 5,
            marginTop: 5,
            marginRight: 5,
            backgroundColor: 'red',
            alignSelf: 'flex-end',
        },
    })

    return (
        <View style={styles.container}>
            <Pressable
                android_ripple={{ color: colors.textPrimary }}
                onPress={() => {
                    onPress(_id)
                }}
                testID="game-card-pressable">
                {teamOneActive && (
                    <Animated.View
                        style={[styles.circle, { opacity: liveOpacity }]}
                    />
                )}
                <View style={styles.teamContainer}>
                    <View style={styles.teamNameContainer}>
                        <Text style={styles.teamText}>{teamOne?.name}</Text>
                        {teamOne?.teamname && (
                            <Text style={styles.teamNameText}>
                                @{teamOne?.teamname}
                            </Text>
                        )}
                    </View>
                    <Text style={styles.teamText}>{teamOneScore}</Text>
                </View>
                <View style={styles.teamContainer}>
                    <View style={styles.teamNameContainer}>
                        <Text style={styles.teamText}>{teamTwo?.name}</Text>
                        {teamTwo?.teamname && (
                            <Text style={styles.teamNameText}>
                                @{teamTwo?.teamname}
                            </Text>
                        )}
                    </View>

                    <Text style={styles.teamText}>{teamTwoScore}</Text>
                </View>
                <View style={styles.footer}>
                    <View style={styles.viewsContainer}>
                        {totalViews > 0 && (
                            <View
                                style={[
                                    styles.viewsContainer,
                                    styles.viewsMargin,
                                ]}>
                                <Text style={styles.footerText}>
                                    {totalViews}
                                </Text>
                                <Icon
                                    style={styles.viewsMargin}
                                    size={20}
                                    name="eye-outline"
                                    color={colors.textPrimary}
                                />
                            </View>
                        )}
                    </View>

                    <View style={styles.scoreLimitContainer}>
                        {teamOneActive && (
                            <Text style={styles.footerText}>
                                Game to {scoreLimit}
                            </Text>
                        )}
                        <IconButton
                            iconColor={colors.textPrimary}
                            icon="chevron-right"
                            testID="go-button"
                        />
                    </View>
                </View>
            </Pressable>
        </View>
    )
}

export default GameCard
