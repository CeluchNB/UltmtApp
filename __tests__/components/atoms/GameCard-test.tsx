import React from 'react'
import GameCard, { GameCardProps } from '../../../src/components/atoms/GameCard'
import { fireEvent, render } from '@testing-library/react-native'
jest.mock('react-native/Libraries/Animated/NativeAnimatedHelper')

let props: GameCardProps
const fn = jest.fn()

beforeEach(() => {
    props = {
        game: {
            _id: 'game1',
            teamOne: {
                _id: 'team1',
                place: 'Place 1',
                name: 'Name 1',
                teamname: 'placename1',
                seasonStart: '2022',
                seasonEnd: '2022',
            },
            teamTwo: {
                _id: 'team2',
                place: 'Place 2',
                name: 'Name 2',
                teamname: 'placename2',
            },
            teamOneScore: 10,
            teamTwoScore: 7,
            scoreLimit: 15,
            teamOneActive: true,
            creator: {} as any,
            teamTwoDefined: false,
            halfScore: 8,
            startTime: new Date('2022'),
            softcapMins: 90,
            hardcapMins: 105,
            playersPerPoint: 7,
            timeoutPerHalf: 1,
            floaterTimeout: true,
            teamTwoActive: false,
            teamOnePlayers: [],
            teamTwoPlayers: [],
            resolveCode: '111111',
            totalViews: 0,
            points: [],
        },
        onPress: fn,
    }
})

it('should match snapshot', () => {
    const snapshot = render(<GameCard {...props} />)

    expect(snapshot.toJSON()).toMatchSnapshot()
})

it('should press', () => {
    const { getByTestId } = render(<GameCard {...props} />)
    const pressable = getByTestId('game-card-pressable')
    fireEvent.press(pressable)
    expect(fn).toHaveBeenCalled()
})
