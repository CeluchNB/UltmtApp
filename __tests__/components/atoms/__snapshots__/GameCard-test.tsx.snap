// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<View
  style={
    {
      "backgroundColor": "#303030",
      "borderColor": "#ffd600",
      "borderRadius": 8,
      "borderWidth": 1,
      "color": "#ffd600",
      "elevation": 10,
      "marginLeft": 5,
      "marginRight": 5,
      "marginTop": 10,
      "shadowColor": "#ffd600",
    }
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    testID="game-card-pressable"
  >
    <View
      collapsable={false}
      style={
        {
          "alignSelf": "flex-end",
          "backgroundColor": "red",
          "borderRadius": 5,
          "height": 10,
          "marginRight": 5,
          "marginTop": 5,
          "opacity": 0,
          "width": 10,
        }
      }
    />
    <View
      style={
        {
          "color": "#ffd600",
          "display": "flex",
          "flexDirection": "row",
          "paddingLeft": 5,
          "paddingRight": 5,
          "paddingTop": 5,
        }
      }
    >
      <View
        style={
          {
            "flex": 1,
          }
        }
      >
        <Text
          style={
            {
              "color": "#7AC6FF",
              "fontSize": 20,
              "fontWeight": "bold",
            }
          }
        >
          Name 1
        </Text>
        <Text
          style={
            {
              "color": "#ffd600",
            }
          }
        >
          @
          placename1
        </Text>
      </View>
      <Text
        style={
          {
            "color": "#7AC6FF",
            "fontSize": 20,
            "fontWeight": "bold",
          }
        }
      >
        10
      </Text>
    </View>
    <View
      style={
        {
          "color": "#ffd600",
          "display": "flex",
          "flexDirection": "row",
          "paddingLeft": 5,
          "paddingRight": 5,
          "paddingTop": 5,
        }
      }
    >
      <View
        style={
          {
            "flex": 1,
          }
        }
      >
        <Text
          style={
            {
              "color": "#7AC6FF",
              "fontSize": 20,
              "fontWeight": "bold",
            }
          }
        >
          Name 2
        </Text>
        <Text
          style={
            {
              "color": "#ffd600",
            }
          }
        >
          @
          placename2
        </Text>
      </View>
      <Text
        style={
          {
            "color": "#7AC6FF",
            "fontSize": 20,
            "fontWeight": "bold",
          }
        }
      >
        7
      </Text>
    </View>
    <View
      style={
        {
          "alignSelf": "flex-end",
          "display": "flex",
          "flexDirection": "row",
        }
      }
    >
      <Text
        style={
          {
            "alignSelf": "center",
            "color": "#ffd600",
            "fontSize": 16,
            "fontWeight": "600",
          }
        }
      >
        Game to 
        15
      </Text>
      <View
        collapsable={false}
        style={
          {
            "backgroundColor": "transparent",
            "borderRadius": 20,
            "height": 40,
            "margin": 6,
            "shadowColor": "#000",
            "shadowOffset": {
              "height": 0,
              "width": 0,
            },
            "shadowOpacity": 0,
            "shadowRadius": 0,
            "width": 40,
          }
        }
        testID="go-button-container-outer-layer"
      >
        <View
          collapsable={false}
          style={
            {
              "backgroundColor": "transparent",
              "borderColor": "rgba(121, 116, 126, 1)",
              "borderRadius": 20,
              "borderWidth": 0,
              "elevation": 0,
              "flex": 1,
              "overflow": "hidden",
              "shadowColor": "#000",
              "shadowOffset": {
                "height": 0,
                "width": 0,
              },
              "shadowOpacity": 0,
              "shadowRadius": 0,
            }
          }
          testID="go-button-container"
        >
          <View
            accessibilityComponentType="button"
            accessibilityRole="button"
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": true,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityTraits="button"
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            centered={true}
            collapsable={false}
            focusable={true}
            hitSlop={
              {
                "bottom": 6,
                "left": 6,
                "right": 6,
                "top": 6,
              }
            }
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              [
                {
                  "overflow": "hidden",
                },
                [
                  {
                    "alignItems": "center",
                    "flexGrow": 1,
                    "justifyContent": "center",
                  },
                  {
                    "borderRadius": 20,
                  },
                ],
              ]
            }
            testID="go-button"
          >
            <Text
              accessibilityElementsHidden={true}
              allowFontScaling={false}
              importantForAccessibility="no-hide-descendants"
              pointerEvents="none"
              selectable={false}
              style={
                [
                  {
                    "color": "#ffd600",
                    "fontSize": 24,
                  },
                  [
                    {
                      "lineHeight": 24,
                      "transform": [
                        {
                          "scaleX": 1,
                        },
                      ],
                    },
                    {
                      "backgroundColor": "transparent",
                    },
                  ],
                  {
                    "fontFamily": "Material Design Icons",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  {},
                ]
              }
            >
              ó°…‚
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
